compiler/src/vm/heap/mod.rs

```diff
@@ -0,0 +1,235 @@
+mod object;
+mod pointer;
+
+pub use self::{
+    object::{Builtin, Closure, Data, Int, Object, Struct, Symbol, Text},
+    pointer::Pointer,
+};
+use crate::builtin_functions::BuiltinFunction;
+use itertools::Itertools;
+use num_bigint::BigInt;
+use std::{cmp::Ordering, collections::HashMap};
+
+#[derive(Clone)]
+pub struct Heap {
+    objects: HashMap<Pointer, Object>,
+    next_address: Pointer,
+}
+
+impl std::fmt::Debug for Heap {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut objects = self.objects.clone().into_iter().collect_vec();
+        objects.sort_by_key(|(address, _)| address.raw());
+
+        writeln!(f, "{{")?;
+        for (address, object) in objects {
+            writeln!(
+                f,
+                "{address}: {} {}",
+                object.reference_count,
+                address.format(self),
+            )?;
+        }
+        write!(f, "}}")
+    }
+}
+
+impl Default for Heap {
+    fn default() -> Self {
+        Self {
+            objects: HashMap::new(),
+            next_address: Pointer::from_raw(1),
+        }
+    }
+}
+impl Heap {
+    pub fn get(&self, address: Pointer) -> &Object {
+        self.objects
+            .get(&address)
+            .unwrap_or_else(|| panic!("Couldn't get object {address}."))
+    }
+    pub fn get_mut(&mut self, address: Pointer) -> &mut Object {
+        self.objects
+            .get_mut(&address)
+            .unwrap_or_else(|| panic!("Couldn't get object {address}."))
+    }
+
+    pub fn dup(&mut self, address: Pointer) {
+        self.get_mut(address).reference_count += 1;
+        log::trace!(
+            "RefCount of {address} increased to {}. Value: {}",
+            self.get(address).reference_count,
+            address.format(self),
+        );
```

Comment:
We can still reduce one lookup here:

```suggestion
        let object = self.get(address);
        trace!(
            "RefCount of {address} increased to {}. Value: {}",
            object.reference_count,
            object.format(self),
        );
```